protocol_connect.go

```diff
@@ -803,48 +794,89 @@ func (u *connectUnaryUnmarshaler) UnmarshalFunc(message any, unmarshal func([]by
 	return nil
 }
 
-type connectWireError Error
+type connectWireDetail ErrorDetail
 
-func (e *connectWireError) MarshalJSON() ([]byte, error) {
-	wire := &errorv1.Error{
-		Code:    CodeUnknown.String(),
-		Message: (*Error)(e).Error(),
+func (d *connectWireDetail) MarshalJSON() ([]byte, error) {
+	if d.wireJSON != "" {
+		// If we unmarshaled this detail from JSON, return the original data. This
+		// lets proxies w/o protobuf descriptors preserve human-readable details.
+		return []byte(d.wireJSON), nil
 	}
-	if connectErr, ok := asError((*Error)(e)); ok {
-		wire.Code = connectErr.Code().String()
-		wire.Message = connectErr.Message()
-		details, err := connectErr.detailsAsAny()
-		if err != nil {
-			return nil, err
-		}
-		wire.Details = details
+	var codec protoJSONCodec
+	out, err := codec.Marshal(d.pb)
+	if err != nil { // likely no descriptor available, fall back to simpler encoding
+		// Protobuf names are ASCII letters, digits, underscore, and period, none
+		// of which need JSON escaping.
+		out = []byte(`{"@type": "`)
+		out = append(out, strings.TrimPrefix(defaultAnyResolverPrefix, d.pb.TypeUrl)...)
+		out = append(out, '"')
 	}
-	return (&protoJSONCodec{}).Marshal(wire)
+	// Since we know that the "@value" key and base64-encoded bytes don't need
+	// escaping, it's simplest to work with the output as raw bytes.
+	out = bytes.TrimRight(out, "}")
+	out = append(out, `, "@value": "`...)
```

Comment:
Are there any edge cases here where someone could break things - i.e. setting `json_name` to `@value`? It might be safest to marshal the JSON representation to another key?

Reply:
I don't think it's possible to generate conflicts here; `@` isn't allowed in protobuf field names, and none of the implementations I'm aware of allow customizing this.

Reply:
Update, I'm uninformed: the `json_name` option allows all sorts of nonsense.

Reply:
Changed to a nested object :/

Reply:
_All_ sorts of nonsense indeed.

```proto
    string a = 3 [json_name = "sameJsonName"];
    string b = 4 [json_name = "sameJsonName"];
    string c = 5;
    string d = 6 [json_name = "c"];
    string e = 7 [json_name = ""];
    string f = 8 [json_name = "@type"];
```

I really liked `@value` and don't think it's a real problem, considering the example above compiles. If `@value` is the source of truth, and the other properties are informal, it would be fine to override any user-defined `@value` with our property.

That said, the nested object does make it clear that the encoded form is the source of truth, and that `debug` is really only there to have a human-readable form on the wire.